### PR TITLE
Evd.Filter - Represent with bitsets

### PR DESCRIPTION
**Pull request by rgrinberg**
Harmless and mostly useless, but I couldn't resist as bitsets done with linked lists just look wrong. I vendored in @backtracking's bitset module (plus a few hastily written helpers), but it can also be done with zarith with a little more work if required.

### CODE REVIEW

File: engine/evd.mli [hunk 1]
```diff
@@ -59,6 +59,8 @@ sig
   (** Horizontal composition : [compose f1 f2] only keeps parts of [f2] where
       [f1] is set. In particular, [f1] and [f2] must have the same length. *)
 
+  val and_ : t -> t -> t
```
Comment (SkySkimmer) [inline engine/evd.mli:62]: If this is the intersection you can call it `inter` without colliding with ocaml keywords.

File: clib/bitv_string.mli [hunk 2]
```diff
@@ -0,0 +1,76 @@
+(**************************************************************************)
+(*                                                                        *)
+(*  Copyright (C) Jean-Christophe Filliatre                               *)
```
Comment (SkySkimmer) [inline clib/bitv_string.mli:3]: This header should also be in the .ml

**Comment by SkySkimmer** [comment]
bench https://gitlab.com/coq/coq/-/jobs/1850923051

File: clib/bitv_string.mli (hunk 2 above)
Comment (backtracking) [reply clib/bitv_string.mli:3]: You're right. Thanks for pointing this out. Now fixed in the `bitv` repo.

**Review by ppedrot — APPROVED** [review]
> Harmless

@rgrinberg I don't think so. Actually the important operation here is that extending a filter is O(1) and shares memory. Otherwise you get quadratic behaviours & memory explosion on big contexts... I don't know if we test this in the bench but I distinctly remember trying bitsets and abandoning the idea on some examples from @JasonGross.

**Review by ppedrot — CHANGES_REQUESTED** [review]
(For some reason I misclicked, this is a comment *against* merging this PR.)

**Comment by ppedrot** [comment]
More generally, any datatype which follows a context structure should never be represented as some variants of arrays. The two most important operations we do with them are cons, fold and sometimes access.

When fast access is not needed, lists are probably both the most natural and best structure experimentally, and when fast access is important, we've been using variants of Okasaki's binary skewed lists which is O(1) cons and O(log n) access.

**Comment by rgrinberg** [comment]
> I don't think so. Actually the important operation here is that extending a filter is O(1) and shares memory. 

Thanks for the pointer and the results confirm it's a wash.

Regarding memory sharing, it doesn't seem like this PR uses more memory. That makes intuitive sense, you need quite a lot of sharing to make up for such an inefficient representation.

However, it's still possible to improve this PR if anyone is interested. The implementation of `extend` really suffers because of my poorly written `Bitv_string.append`. A proper implementation should be like an order of magnitude faster. It looks like @backtracking hasn't gotten around to it yet. Perhaps it will be worth trying again with a fast append.

**Comment by backtracking** [comment]
For what is worth, I'm mentioning the existence of this implementation of flexible arrays:
https://github.com/backtracking/flex-array

**Comment by rgrinberg** [comment]
> For what is worth, I'm mentioning the existence of this implementation of flexible arrays:

Cool. I'm not sure how useful it is in this case though. It's mostly the equivalent of `List.cons true` that needs optimizing, so perhaps they could represent the filter as:

```ocaml
type t =
  | Vector of { true_prefix : int ; suffix : bool list }
  | Identity
```

To cut down on the memory junk but still keep the fast `cons true`.

**Pull request closed without merging**